### PR TITLE
dm: Improve the mpsl timeslot length setting

### PR DESCRIPTION
**Pull request by maje-emb**
The timeslot length and timeout for the nrf_dm_proc_execute() function
is calculated using the nrf_dm_get_duration_us() function.
In addition, these times can be extended by an additional
ranging window time, which is configured in run-time.

### CODE REVIEW

File: subsys/dm/dm.c [hunk 1]
```diff
@@ -115,36 +119,44 @@ static mpsl_timeslot_request_t timeslot_request_normal = {
...
-		dm_config.ranging_mode = NRF_DM_RANGING_MODE_MCPD;
+	if (dm_req->ranging_mode == DM_RANGING_MODE_RTT) {
+		dm_config->ranging_mode = NRF_DM_RANGING_MODE_RTT;
 	} else {
-		return NRF_DM_STATUS_CONF_INVALID;
+		dm_config->ranging_mode = NRF_DM_RANGING_MODE_MCPD;
+	}
+}
+
+static uint32_t dm_proc_execute_duration(struct dm_request *dm_req)
```
Comment (eriksandgren) [inline subsys/dm/dm.c:145]:
```suggestion
static uint32_t dm_proc_execute_duration_us(struct dm_request *dm_req)
```

File: subsys/dm/dm.c [hunk 2]
```diff
@@ -115,36 +119,44 @@ static mpsl_timeslot_request_t timeslot_request_normal = {
...
+{
+	uint32_t time;
```
Comment (eriksandgren) [inline subsys/dm/dm.c:147]:
```suggestion
	uint32_t time_us;
```

File: subsys/dm/dm.c (hunk 1 above)
Comment (eriksandgren) [reply subsys/dm/dm.c:145]: I see that in many other places when dealing with timing calculations _us is being used to indicate the unit

File: subsys/dm/dm.c [hunk 3]
```diff
@@ -167,12 +180,11 @@ static mpsl_timeslot_signal_return_param_t *mpsl_timeslot_callback(
...
-		nrf_dm_status = dm_configure();
+		dm_config_get(&timeslot_ctx.curr_req.dm_req, &dm_config);
+		nrf_dm_status = nrf_dm_configure(&dm_config);
 
 		if (nrf_dm_status == NRF_DM_STATUS_SUCCESS) {
-			uint32_t timeout = timeslot_ctx.curr_req.dm_req.role == DM_ROLE_INITIATOR ?
-					 INITIATOR_RANGING_WINDOW_US : REFLECTOR_RANGING_WINDOW_US;
-			nrf_dm_status = nrf_dm_proc_execute(timeout);
+			nrf_dm_status = nrf_dm_proc_execute(timeslot_ctx.curr_req.window_length);
 		}
```
Comment (eriksandgren) [inline subsys/dm/dm.c:188]: I realise this is unrelated to you current PR, but if the  `nrf_dm_configure` call does not return `NRF_DM_STATUS_SUCCESS` then some invalid parameters are likely being set in the `dm_config`. Maybe this should be logged somehow.

File: subsys/dm/dm.c [hunk 4]
```diff
@@ -355,12 +368,18 @@ static void dm_start_ranging(void)
 
 static void dm_reschedule(void)
 {
+	uint32_t timeslot_len;
```
Comment (eriksandgren) [inline subsys/dm/dm.c:371]:
```suggestion
	uint32_t timeslot_len_us;
```

File: subsys/dm/dm.c [hunk 5]
```diff
@@ -355,12 +368,18 @@ static void dm_start_ranging(void)
...
+	uint32_t window_len;
```
Comment (eriksandgren) [inline subsys/dm/dm.c:372]:
```suggestion
	uint32_t window_len_us;
```

File: subsys/dm/dm.c [hunk 6]
```diff
@@ -435,15 +454,17 @@ static void dm_thread(void)
 int dm_request_add(struct dm_request *req)
 {
 	int err;
+	uint32_t timeslot_len;
```
Comment (eriksandgren) [inline subsys/dm/dm.c:457]:
```suggestion
	uint32_t timeslot_len_us;
```

File: subsys/dm/dm.c [hunk 7]
```diff
@@ -435,15 +454,17 @@ static void dm_thread(void)
...
+	uint32_t window_len;
```
Comment (eriksandgren) [inline subsys/dm/dm.c:458]:
```suggestion
	uint32_t window_len_us;
```

File: subsys/dm/timeslot_queue.h [hunk 8]
```diff
@@ -14,32 +14,34 @@
...
 /** @brief Timeslot request structure */
 struct timeslot_request {
 	/* Distance measurement request structure */
 	struct dm_request dm_req;
 
 	/* The desired start time of timeslot */
 	uint32_t start_time;
+
+	/* Timeslot length*/
+	uint32_t timeslot_length;
```
Comment (eriksandgren) [inline subsys/dm/timeslot_queue.h:26]:
```suggestion
	uint32_t timeslot_length_us;
```

File: subsys/dm/timeslot_queue.h [hunk 9]
```diff
@@ -14,32 +14,34 @@
...
+
+	/* Ranging window length */
+	uint32_t window_length;
```
Comment (eriksandgren) [inline subsys/dm/timeslot_queue.h:29]:
```suggestion
	uint32_t window_length_us;
```

File: subsys/dm/dm.c [hunk 10]
```diff
@@ -60,6 +60,10 @@ LOG_MODULE_REGISTER(nrf_dm, CONFIG_DM_MODULE_LOG_LEVEL);
 #define DM_THREAD_PRIORITY           K_HIGHEST_APPLICATION_THREAD_PRIO
 #define TIMESLOT_TIMEOUT_STEP_MS     120000
 
+#define DM_TIMESLOT_OVERHEAD_US      400
+#define DM_REFLECTOR_OVERHEAD_US     2000
+#define DM_RANGING_EXTRA_TIMEOUT_US  200
```
Comment (eriksandgren) [inline subsys/dm/dm.c:65]: I guess this one will be removed once we have the fix from nrf_lib_dm library. I could provide you a lib that includes the fix if you want to test.

**Review by eriksandgren — CHANGES_REQUESTED** [review]
Some suggestions to add the _us postfix to variables that represents a duration in microseconds. I might have missed some places where this can be changed.

File: subsys/dm/dm.c (hunk 1 above)
Comment (maje-emb) [reply subsys/dm/dm.c]: Good point. Added the _us prefix.

File: subsys/dm/dm.c (hunk 3 above)
Comment (maje-emb) [reply subsys/dm/dm.c:188]: Added log message, if the returned value is not `NRF_DM_STATUS_SUCCESS`

File: subsys/dm/dm.c (hunk 10 above)
Comment (maje-emb) [reply subsys/dm/dm.c:65]: Yes, `DM_RANGING_EXTRA_TIMEOUT_US` will be removed when we have the fix in nrf_lib_dm.

**Review by eriksandgren — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst [hunk 11]
```diff
@@ -487,6 +487,10 @@ Other libraries
   * Changed the library implementation to bypass the flash driver when storing the emergency data.
     This allows calling the :c:func:`emds_store` function from an interrupt context.
 
+* :ref:`mod_dm`:
+  * Added run-time ranging window length configuration.
+  * Improved the calculation of the mpsl timeslot length by using the functionality of the :ref:`nrf_dm` library.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:492]:
```suggestion
  * Added window length configuration for the runtime ranging window.
  * Improved the calculation of MPSL timeslot length by using the :ref:`nrf_dm` library functionality.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: doc/nrf/releases/release-notes-changelog.rst (hunk 11 above)
Comment (eriksandgren) [reply doc/nrf/releases/release-notes-changelog.rst:492]: I think the point with original sentence is that the "ranging window" is configurable at runtime now, there is no "runtime ranging window" :nerd_face:

Comment (maje-emb) [reply doc/nrf/releases/release-notes-changelog.rst:492]: Maybe something like this:
_Added window length configuration in run-time, when a new measurement request is added._

Comment (peknis) [reply doc/nrf/releases/release-notes-changelog.rst:492]: Right, my bad. I just had an issue with the long list of nouns, not really understanding what defines what.

Would this be clear: "Added a window length configuration to be used runtime, when a new measurement request is added."?

**Review by peknis — APPROVED** [review]

**Review by Tschet1 — APPROVED** [review]

**Pull request merged**